stripe.go

```diff
@@ -621,11 +631,37 @@ func (s *BackendImplementation) shouldRetry(err error, req *http.Request, resp *
 
 	stripeErr, _ := err.(*Error)
 
-	// TODO: This retries any non-Stripe errors produced as part of the
-	// communication process. It generally works, but there are many errors
-	// that should *not* be retried. Try to make this more granular by
-	// including only connection errors, timeout errors, etc.
+	// Don't retry if the context was was cancelled or its deadline was
+	// exceeded.
+	if req.Context() != nil && req.Context().Err() != nil {
+		return false
+	}
+
+	// We retry most errors that come out of HTTP requests except for a curated
+	// list that we know not to be retryable. This list is probably not
+	// exhaustive, so it'd be okay to add new errors to it. It'd also be okay to
+	// flip this to an inverted strategy of retrying only errors that we know
+	// to be retryable in a future refactor, if a good methodology is found for
+	// identifying that full set of errors.
 	if stripeErr == nil && err != nil {
+		if urlErr, ok := err.(*url.Error); ok {
+			// Don't retry too many redirects.
+			if redirectsErrorRE.MatchString(urlErr.Error()) {
+				return false
+			}
+
+			// Don't retry invalid protocol scheme.
+			if schemeErrorRE.MatchString(urlErr.Error()) {
+				return false
+			}
+
+			// Don't retry TLS certificate validation problems.
+			if _, ok := urlErr.Err.(x509.UnknownAuthorityError); ok {
+				return false
+			}
```

Comment:
I don't know much about Stripe's HTTP frontend infrastructure, but I wonder if this case should be omitted for your use-case.

In the exceedingly rare case that Stripe rolled out a bad configuration that included bad TLS certificates, wouldn't there be a chance that the retry is routed to a healthy server, thus recovering the transaction?

Reply:
I think it's okay to leave — it'd be a massive catastrophe if we deployed an invalid TLS certificate, so what this is more likely to protect against is users who have misconfigured their DNS or proxy, and are hitting a non-Stripe server instead of Stripe. We've seen this kind of issue once in a while in the past.